### PR TITLE
Doc/v3 migration guide

### PR DESCRIPTION
**Pull request by jhamman**
Over the past few weeks, we've had a number of conversations/questions about the policy for backward compatibility, deprecations, and breaking changes with the upcoming 3.0 release. This doc is meant to help us iterate toward common language. In its initial form, it is not complete.

Goals for the text here:

1. Developers of 2.18 and 3.0 should be able to decide if backward compatibility is a required attribute of a contribution
2. Users of Zarr should be able to understand if their application will be impacted by the upcoming 3.0 release
3. Users of Zarr should be able to make a plan for how they will adapt their usage of Zarr after the release
1. [non-goal] This is not meant to provide a comprehensive listing of the changes to the zarr API 

cc @zarr-developers/python-core-devs

### CODE REVIEW

**Comment by TomAugspurger** [comment]
One thing that might be helpful: what's the group's tolerance for either compatibility code or deprecations as a way to ease the transition? It sounds like strict backwards compatibility (perhaps with warnings) isn't a goal. Is there tolerance for things like https://github.com/zarr-developers/zarr-python/pull/2098 (e.g. restore some properties to the `Group` object, loosen the keyword-only requirement for some functions). Likewise for things like "cleaning up internal and user facing APIs", which could be done with a  deprecation warning. Even if there isn't tolerance for backwards compatibility shims that really clash with the V3 spec or the current v3 implementation?

File: docs/migration.rst [hunk 1]
```diff
@@ -0,0 +1,89 @@
+Zarr-Python 3 Migration Guide
+=============================
+
+Zarr-Python 3 introduces a number of changes to the API, including a number
+of significant breaking changes and pending deprecations.
+
+This page provides a guide highlighting the most notable changes to help you
+migrate your code from Zarr-Python 2.x to Zarr-Python 3.x.
```
Comment (dstansby) [inline docs/migration.rst:8]:
```suggestion
migrate your code from version 2 to version 3.
```

File: docs/migration.rst [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+Compatibility target
+--------------------
+
+Zarr-Python 3 represents a major refactor of the Zarr-Python codebase. Some of the goals motivating this refactor included:
+
+  - adding support for the V3 specification (alongside the V3 specification)
```
Comment (dstansby) [inline docs/migration.rst:15]:
```suggestion
  - adding support for the Zarr V3 specification (alongside the Zarr V2 specification)
```

File: docs/migration.rst [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  - cleaning up internal and user facing APIs
+  - improving performance (particularly in high latency storage environments like cloud object store)
+
+Though these goals necessitated some breaking changes to the API (hence the major version update), we have tried to maintain
```
Comment (dstansby) [inline docs/migration.rst:19]:
```suggestion
These goals necessitated some breaking changes to the API (hence the major version update), but we have tried to maintain
```

File: docs/migration.rst [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+backwards compatibility in the most widely used parts of the API including the `Array` and `Group` classes and the top-level
+API (e.g. `zarr.open_array` and `zarr.open_group`). It is worth noting that we significantly evolved the internal data model,
+moving away from a model that was tightly coupled to the v2 specification, and to a more generic representation of Zarr objects.
```
Comment (dstansby) [inline docs/migration.rst:22]: This sentence raises more questions (why did we evolve the data model? how did we evolve the data model? what does the new model look like?), and I don't think is needed for a simple compatability guide, so I think best to drop?
```suggestion
API (e.g. `zarr.open_array` and `zarr.open_group`).
```

File: docs/migration.rst [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+Getting ready for 3.0
+---------------------
+
+Ahead of the 3.0 release, we suggest projects that depend on Zarr-Python take the following actions:
+
+1. Pin the supported Zarr-Python version to ``zarr>=2,<3``. This is a best practice and will protect your users from any incompatibilities that may arise during the release of Zarr-Python 3.0.
+2. Limit your imports from the Zarr-Python package. Most of the primary API ``zarr.*`` will be compatible in 3.0. However, the following breaking API changes are planned:
```
Comment (dstansby) [inline docs/migration.rst:30]: Shouldn't this list have

3. Test that your package works with v3
4. Update the pin to zarr >=3

At the end?

File: docs/migration.rst [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+   
+   - ``numcodecs.*`` will no longer be available in ``zarr.*``. (Suggested action: transition to importing codecs from ``numcodecs`` directly.)
```
Comment (dstansby) [inline docs/migration.rst:32]:
```suggestion
   - ``zarr.numcodecs.*`` will no longer be available. These imports can be replaced by importing ``numcodecs`` directly.
```
We should be doing more than "suggesting", we should be providing concrete fixes/code updates!

File: docs/migration.rst [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+   - The ``zarr.v3_api_available`` feature flag is being removed. (Suggested action: this experimental feature was deprecated in v2.18.)
```
Comment (dstansby) [inline docs/migration.rst:33]:
```suggestion
   - The ``zarr.v3_api_available`` feature flag is being removed. In zarr-python v3 the v3 API is always available, so you shouldn't need to use this flag.
```

File: docs/migration.rst [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    - ``zarr.core``
+    - ``zarr.hierarchy``
+    - ``zarr.indexing``
+    - ``zarr.meta``
+    - ``zarr.meta_v1``
+    - ``zarr.storage``
+    - ``zarr.sync``
+    - ``zarr.types``
+    - ``zarr.util``
+    - ``zarr.n5``
```
Comment (dstansby) [inline docs/migration.rst:48]: This needs a clear list of how to update code to adapt to these removals or changes.

File: docs/migration.rst [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+Continue using Zarr-Python 2
+----------------------------
+
+Zarr-Python 2.x is still available, though we recommend migrating to Zarr-Python 3 for its improvements and new features.
+Security and bug fixes will be made to the 2.x series for at least 6 months following the first Zarr-Python 3 release.
+
+If you need to use the latest Zarr-Python 2 release, you can install it with:
+
+    $ pip install "zarr==2.*"
```
Comment (dstansby) [inline docs/migration.rst:58]: I don't think this needs a whole section - perhaps just mention zarr-python 2 is available for at least 6 months above somewhere?

**Review by dstansby — CHANGES_REQUESTED** [review]
I left some comments - I'll mark this as request changes since it still needs major additions.

File: docs/migration.rst (hunk 6 above)
Comment (jhamman) [reply docs/migration.rst:32]: I wish it were that easy! Unfortunately, the top level namespace is littered with `*` imports

https://github.com/zarr-developers/zarr-python/blob/e1d98cdf3d35e4cb2815912d82a64c1623347c87/zarr/__init__.py#L2
https://github.com/zarr-developers/zarr-python/blob/e1d98cdf3d35e4cb2815912d82a64c1623347c87/zarr/codecs.py#L2

If there is a reasonable way for us to deprecate folks using these imports, I'm all ears.

File: docs/migration.rst (hunk 9 above)
Comment (jhamman) [reply docs/migration.rst:58]: I see what you are saying but this has come up a lot. In this case, I'd like to be clear, even if it is obvious.

File: docs/migration.rst [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    - ``zarr.hierarchy``
+    - ``zarr.indexing``
+    - ``zarr.meta``
+    - ``zarr.meta_v1``
+    - ``zarr.storage``
+    - ``zarr.sync``
+    - ``zarr.types``
+    - ``zarr.util``
+    - ``zarr.n5``
+
```
Comment (jhamman) [inline docs/migration.rst:49]:
```suggestion

3. Test that your package works with v3. You can start testing against version 3 now (pre-releases are being published to PyPI weekly).
4. Update the pin to zarr >=3
```

File: docs/migration.rst [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+---------------------
+
+Ahead of the 3.0 release, we suggest projects that depend on Zarr-Python take the following actions:
+
+1. Pin the supported Zarr-Python version to ``zarr>=2,<3``. This is a best practice and will protect your users from any incompatibilities that may arise during the release of Zarr-Python 3.0.
+2. Limit your imports from the Zarr-Python package. Most of the primary API ``zarr.*`` will be compatible in 3.0. However, the following breaking API changes are planned:
+   
+   - ``numcodecs.*`` will no longer be available in ``zarr.*``. (Suggested action: transition to importing codecs from ``numcodecs`` directly.)
+   - The ``zarr.v3_api_available`` feature flag is being removed. (Suggested action: this experimental feature was deprecated in v2.18.)
+   - The following internal modules are being removed or significant changed:
```
Comment (jhamman) [inline docs/migration.rst:34]:
```suggestion
   - The following internal modules are being removed or significantly changed. If your application relies on imports from any of the below modules, you will need to either a) modify your application to no longer rely on these imports or b) vendor the parts of the specific modules that you need.
```

File: docs/migration.rst [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    - ``zarr.meta_v1``
+    - ``zarr.storage``
+    - ``zarr.sync``
+    - ``zarr.types``
+    - ``zarr.util``
+    - ``zarr.n5``
+
+3. Test that your package works with v3. You can start testing against version 3 now (pre-releases are being published to PyPI weekly).
+4. Update the pin to zarr >=3
+Continue using Zarr-Python 2
```
Comment (jhamman) [inline docs/migration.rst:51]:
```suggestion

Continue using Zarr-Python 2
```

**Comment by jni** [comment]
I just came here following the link from #1849. Currently, this guide feels very incomplete, and since #2182 is now in-flight, I think this guide should become very clear.

From the perspective of a maintainer / close user of a bunch of libraries that depend on zarr *but have extremely limited maintainer time*, the most important thing I want to understand is how hard it will be to support *both* zarr 2.18 and zarr 3+ within a single library. I think this is *the* critical question for a smooth transition, because it is hard for libraries to all migrate at the same time, and you want libraries to be installable together in the same environment — you don't want someone depending on both napari and ome-zarr to face napari requiring zarr>=3 and ome-zarr requiring zarr<3. So many libraries would want to support the subset of zarr that is identical in v3 and v2 until everyone can agree to depend on 3+.

File: docs/migration.rst (hunk 6 above)
Comment (dstansby) [reply docs/migration.rst:31]: Even if there isn't a reasonable way to deprecate the imports, as a minimum this migration guide should provide a complete list of all imports that are disappearing and how to replace them.

**Comment by jhamman** [comment]
@jni - thanks for the feedback. I agree this is not ready to ship yet. The main things that we know are changing is the `Store` API and access to internal APIs (e.g. `zarr.core.xxx`). Beyond that, the best way for us to fill out the migration guide is to have projects attempt to support v3 and report back. I'm doing that with Dask right now (https://github.com/dask/dask/pull/11388, https://github.com/zarr-developers/zarr-python/pull/2186) and I understand @TomAugspurger has begun the process for Xarray. So an ask for you and the Napari / ome-zarr devs is to try to do this and report back. Beyond that, specific suggestions to this doc are more than welcome (props to @dstansby for his edits already).

**Comment by dstansby** [comment]
If we want downstream packages to test with version 3.0.0a1, it might be good to do a blog post or add something to the docs explaining how to do that testing, what to look for, and how to provide feedback?

**Comment by jni** [comment]
Well, @Czaki started doing that for us in napari/napari#7215 and @d-v-b has [been helping](https://napari.zulipchat.com/#narrow/stream/212875-general/topic/handling.20zarr.203.2E0.20release/near/445118833) 🙏.

It looks like one of the remaining issues is that `zarr.open` defaults to v3 zarr (it'd be worth considering switching to calver, which I hate, *if only* to avoid the confusion between zarr format v3 and zarr-python v3... 😂), and tensorstore does not yet support v3 zarr files. At least that's my interpretation of [these lines](https://github.com/napari/napari/pull/7215/files#diff-7e3cbb11d497f0a4fe3344e5d99e8453b9037f5a073bc5642e00ebca69af6316R1119). Nor do I see any motions to change this in the tensorstore repo... @jbms?

But I think we can resolve this by explicitly writing a v2 zarr in the test?

> specific suggestions to this doc are more than welcome

Something along the lines of:

> Common functions have switched to keyword-only arguments, so you will need to change any invocation of, for example, `zarr.open(path, 'a')` to `zarr.open(store=path, mode='a')`.

(An exhaustive list of such changes would be useful.)

**Comment by d-v-b** [comment]
> and tensorstore does not yet support v3 zarr files.

Tensorstore has support zarr v3 for a long time: https://google.github.io/tensorstore/driver/zarr3/index.html

**Comment by jni** [comment]
> Tensorstore has supported zarr v3 for a long time

oh, interesting, thanks for pointing that out @d-v-b! I couldn't actually find the relevant PR but did not look exhaustively. 🙏

**Comment by TomAugspurger** [comment]
A few changes I've found while updating xarray. Are all of these intentional?

- `Array.resize` returns a new `Array` object. 2.x mutated the `Array` in place
- `zarr_version` has been renamed to `zarr_format`
- Some exception types have changed (e.g. 2.x raised a `zarr.errors.GroupNotFoundError` while 3.x raises a `ValueError`)
- `write_empty_chunks` has been removed

**Pull request closed without merging**